source/server/admin/config_dump_handler.cc

```diff
@@ -176,7 +180,14 @@ ConfigDumpHandler::addResourceToDump(envoy::admin::v3::ConfigDump& dump,
       if (mask.has_value()) {
         Protobuf::FieldMask field_mask;
         ProtobufUtil::FieldMaskUtil::FromString(mask.value(), &field_mask);
-        trimResourceMessage(field_mask, msg);
+        // trimResourceMessage will LOG(FATAL) if `field_mask` is malformed.
+        TRY_ASSERT_MAIN_THREAD { trimResourceMessage(field_mask, msg); }
+        END_TRY
+        catch (...) {
```

Comment:
Actually, it seems as though this isn't an EnvoyException but instead a std::exception, and calling e.what() just returns `"std::exception"`, which isn't particularly helpful.

Reply:
Can you debug who is throwing this std::exception?

In theory no one in Envoy should be throwing an exception that is not derived from EnvoyException. However, if this comes from the STL then that won't be true.

Reply:
This is coming from a LOG(FATAL) [here](https://github.com/protocolbuffers/protobuf/blob/master/src/google/protobuf/generated_message_reflection.cc#L122).

Reply:
Can we see what's going on up stack and make it not call LOG(FATAL)?

Reply:
Full stack is as below:

```
Protobuf::(anonymous namespace)::ReportReflectionUsageError()
Protobuf::Reflection::HasField()
ProtobufUtil::FieldMaskUtil::(anonymous namespace)::FieldMaskTree::TrimMessage()
ProtobufUtil::FieldMaskUtil::(anonymous namespace)::FieldMaskTree::TrimMessage()
ProtobufUtil::FieldMaskUtil::(anonymous namespace)::FieldMaskTree::TrimMessage()
ProtobufUtil::FieldMaskUtil::FieldMaskUtil::TrimMessage()
Envoy::Server::ConfigDumpHandler::addAllConfigToDump()
Envoy::Server::ConfigDumpHandler::handlerConfigDump()
```
So it seems as though we'd need to check the mask before handing it to TrimMesage in order to avoid the LOG(FATAL), as TrimMessage is already in google/protobuf.

Reply:
@htuch @markdroth is there a gRPC API for detecting an invalid protobuf without crashing?

Reply:
@jmarantz do you mean protobuf library or gRPC? Is the proto invalid or is this some field mask library issue?

Reply:
This is an issue where if the FieldMask is malformed (e.g. attempts to index through a repeated field), the protobuf library crashes via LOG(FATAL). There is a method in FieldMaskUtil to check whether a field mask is valid, but requires the proto to be statically typed, whereas the protos in config_dump are all dynamically typed out of general Protobuf::Message or Any.

Reply:
Can you use this utility inside `trimResourceMessage` after the unpack?

Reply:
I meant protobuf. Thanks.

Reply:
Unfortunately I don't think so; all messages are dynamically constructed from DynamicMessageFactory inside trimResourceMessage, and [IsValidFieldMask](https://developers.google.com/protocol-buffers/docs/reference/cpp/google.protobuf.util.field_mask_util#FieldMaskUtil) is templated and requires compile-time knowledge of the proto type, e.g. `IsValidFieldMask<Cluster>(mask)`.

Reply:
Is there anything we can do to pre-validate the field_mask to avoid crashing in protobuf? Catching the throw that emanates from a LOG(FATAL) seems non-robust.

Reply:
I added a method which essentially does the same thing as `IsValidFieldMask` but doesn't require compile-time knowledge of types. LMK if this works for you!